server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java

```diff
@@ -114,17 +120,120 @@ public String toString() {
         }
     }
 
-    record ClusterFormationState(
-        Settings settings,
-        ClusterState clusterState,
+    /**
+     * If this node believes that cluster formation has failed, this record provides information that can be used to determine why that is.
+     */
+    public record ClusterFormationState(
+        List<String> initialMasterNodesSetting,
+        DiscoveryNode localNode,
+        Map<String, DiscoveryNode> masterEligibleNodes,
+        long clusterStateVersion,
+        long acceptedTerm,
+        VotingConfiguration lastAcceptedConfiguration,
+        VotingConfiguration lastCommittedConfiguration,
         List<TransportAddress> resolvedAddresses,
         List<DiscoveryNode> foundPeers,
         long currentTerm,
-        ElectionStrategy electionStrategy,
+        boolean hasDiscoveredQuorum,
         StatusInfo statusInfo,
         List<JoinStatus> inFlightJoinStatuses
-    ) {
-        String getDescription() {
+    ) implements Writeable {
+
+        public ClusterFormationState(
+            Settings settings,
+            ClusterState clusterState,
+            List<TransportAddress> resolvedAddresses,
+            List<DiscoveryNode> foundPeers,
+            long currentTerm,
+            ElectionStrategy electionStrategy,
+            StatusInfo statusInfo,
+            List<JoinStatus> inFlightJoinStatuses
+        ) {
+            this(
+                INITIAL_MASTER_NODES_SETTING.get(settings),
+                clusterState.nodes().getLocalNode(),
+                clusterState.nodes().getMasterNodes(),
+                clusterState.version(),
+                clusterState.term(),
+                clusterState.getLastAcceptedConfiguration(),
+                clusterState.getLastCommittedConfiguration(),
+                resolvedAddresses,
+                foundPeers,
+                currentTerm,
+                calculateHasDiscoveredQuorum(
+                    foundPeers,
+                    electionStrategy,
+                    clusterState.nodes().getLocalNode(),
+                    currentTerm,
+                    clusterState.term(),
+                    clusterState.version(),
+                    clusterState.getLastCommittedConfiguration(),
+                    clusterState.getLastAcceptedConfiguration()
+                ),
+                statusInfo,
+                inFlightJoinStatuses
+            );
+        }
+
+        private static boolean calculateHasDiscoveredQuorum(
```

Comment:
nit: would it make sense to obey the `isXXX`/`hasXXX` naming convention for methods returning boolean?